python/ray/data/_internal/execution/interfaces/physical_operator.py

```diff
@@ -1,14 +1,114 @@
-from typing import Callable, Dict, List, Optional
+from abc import ABC, abstractmethod
+from typing import Callable, Dict, List, Optional, Union
 
 import ray
 from .ref_bundle import RefBundle
+from ray._raylet import StreamingObjectRefGenerator
 from ray.data._internal.execution.interfaces.execution_options import (
     ExecutionOptions,
     ExecutionResources,
 )
 from ray.data._internal.logical.interfaces import Operator
 from ray.data._internal.stats import StatsDict
 
+# TODO(hchen): Ray Core should have a common interface for these two types.
+Waitable = Union[ray.ObjectRef, StreamingObjectRefGenerator]
+
+
+class OpTask(ABC):
+    """Abstract class that represents a task that is created by an PhysicalOperator.
+
+    The task can be either a regular task or an actor task.
+    """
+
+    @abstractmethod
+    def get_waitable(self) -> Union[ray.ObjectRef, StreamingObjectRefGenerator]:
+        """Return the ObjectRef or StreamingObjectRefGenerator to wait on."""
+        pass
+
+    @abstractmethod
+    def on_waitable_ready(self):
+        """Called when the waitable is ready."""
```

Comment:
This is called multiple times for the streaming refs right? Should clarify that in the comment.